### PR TITLE
Key Connector feature toggle

### PR DESCRIPTION
**Pull request by Hinton**
## Type of change
- [x] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other

## Objective

Since the Key Connector functionality can be tricky to setup and should only be available for certain plans, we've added a feature toggle.

Resolves https://app.asana.com/0/1198901840263430/1201363553288934

## Before you submit
- [x] If making database changes - I have also updated Entity Framework queries and/or migrations
- [ ] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [ ] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)

### CODE REVIEW

File: src/Admin/Models/OrganizationEditModel.cs [hunk 1]
```diff
@@ -78,6 +79,8 @@ public class OrganizationEditModel : OrganizationViewModel
         public bool UsePolicies { get; set; }
         [Display(Name = "SSO")]
         public bool UseSso { get; set; }
+        [Display(Name = "Customer Managed Encryption")]
```
Comment (eliykat) [inline src/Admin/Models/OrganizationEditModel.cs:82]: We use `Key Connector` in the web vault interface, should we use the same wording here for consistency?

File: src/Core/Services/Implementations/OrganizationService.cs [hunk 2]
```diff
@@ -865,6 +877,16 @@ public async Task UpdateLicenseAsync(Guid organizationId, OrganizationLicense li
                 }
             }
 
+            if (!license.UseKeyConnector && organization.UseKeyConnector)
+            {
+                var ssoConfig = await _ssoConfigRepository.GetByOrganizationIdAsync(organization.Id);
+                if (ssoConfig != null && ssoConfig.GetData().UseKeyConnector)
+                {
+                    throw new BadRequestException($"Your organization currently has Key Connector enabled. " +
+                        $"Your new license does not allow for the use of Key Connector. Disable your Key Connector.");
```
Comment (eliykat) [inline src/Core/Services/Implementations/OrganizationService.cs:886]: Seems a bit cheeky telling them to disable something that they can't disable (yet), but it seems like an edge case that we can ignore for now.

File: test/Core.Test/Services/SsoConfigServiceTests.cs [hunk 3]
```diff
@@ -220,12 +225,73 @@ public async Task SaveAsync_KeyConnector_SsoConfigNotEnabled_Throws(SutProvider<
...
+            sutProvider.GetDependency<IPolicyRepository>().GetByOrganizationIdTypeAsync(
+                Arg.Any<Guid>(), Arg.Any<Enums.PolicyType>()).Returns(new Policy
+            {
+                Enabled = true,
+            });
+
+            var exception = await Assert.ThrowsAsync<BadRequestException>(
+                () => sutProvider.Sut.SaveAsync(ssoConfig, organization));
+
+            Assert.Contains("Organization cannot use key connector.", exception.Message);
```
Comment (eliykat) [inline test/Core.Test/Services/SsoConfigServiceTests.cs:262]: Will also need to change this after changing the error message

**Review by eliykat — CHANGES_REQUESTED** [review]

File: src/Core/Services/Implementations/SsoConfigService.cs [hunk 4]
```diff
@@ -63,8 +63,13 @@ private async Task<bool> AnyOrgUserHasKeyConnectorEnabledAsync(Guid organization
             return userDetails.Any(u => u.UsesKeyConnector);
         }
 
-        private async Task VerifyDependenciesAsync(SsoConfig config)
+        private async Task VerifyDependenciesAsync(SsoConfig config, Organization organization)
         {
+            if (!organization.UseKeyConnector)
+            {
+                throw new BadRequestException("Organization cannot use key connector.");
```
[the remark this answers is not in the document]
Comment (Hinton) [reply src/Core/Services/Implementations/SsoConfigService.cs]: The users shouldn't se the error since the UI prevents it.

File: src/Admin/Models/OrganizationEditModel.cs (hunk 1 above)
Comment (Hinton) [reply src/Admin/Models/OrganizationEditModel.cs:82]: Hmm, valid point. My thinking was that the feature is called "Customer Managed Encryption" which might make it more obvious to use that name vs the internal. I'll check with Product.

File: src/Core/Models/Api/Response/ProfileOrganizationResponseModel.cs [hunk 5]
```diff
@@ -14,6 +14,7 @@ public ProfileOrganizationResponseModel(OrganizationUserOrganizationDetails orga
             Name = organization.Name;
             UsePolicies = organization.UsePolicies;
             UseSso = organization.UseSso;
+            UseKeyConnector = organization.UseKeyConnector;
```
[the remark this answers is not in the document]
Comment (Hinton) [reply src/Core/Models/Api/Response/ProfileOrganizationResponseModel.cs:17]: I agree, let's renamed it.

**Comment by eliykat** [comment]
Your changes look good to me, except your tests are failing (maybe related to the error message changing?)

**Review by eliykat — APPROVED** [review]

**Review by eliykat — APPROVED** [review]
